crates/cubecl-hip/src/compute/storage.rs

```diff
@@ -0,0 +1,125 @@
+use cubecl_hip_sys::HIP_SUCCESS;
+use cubecl_runtime::storage::{ComputeStorage, StorageHandle, StorageId, StorageUtilization};
+use std::collections::HashMap;
+
+/// Buffer storage for HIP.
+pub struct HipStorage {
+    memory: HashMap<StorageId, cubecl_hip_sys::hipDeviceptr_t>,
+    deallocations: Vec<StorageId>,
+    stream: cubecl_hip_sys::hipStream_t,
+    activate_slices: HashMap<ActiveResource, cubecl_hip_sys::hipDeviceptr_t>,
+}
+
+#[derive(new, Debug, Hash, PartialEq, Eq, Clone)]
+struct ActiveResource {
+    ptr: u64,
+}
+
+unsafe impl Send for HipStorage {}
+
+impl core::fmt::Debug for HipStorage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(format!("HipStorage {{ device: {:?} }}", self.stream).as_str())
+    }
+}
+
+/// Keeps actual wgpu buffer references in a hashmap with ids as key.
+impl HipStorage {
+    /// Create a new storage on the given [device](wgpu::Device).
+    pub fn new(stream: cubecl_hip_sys::hipStream_t) -> Self {
+        Self {
+            memory: HashMap::new(),
+            deallocations: Vec::new(),
+            stream,
+            activate_slices: HashMap::new(),
+        }
+    }
+
+    /// Actually deallocates buffers tagged to be deallocated.
+    pub fn perform_deallocations(&mut self) {
+        for id in self.deallocations.drain(..) {
+            if let Some(ptr) = self.memory.remove(&id) {
+                unsafe {
+                    cubecl_hip_sys::hipFreeAsync(ptr, self.stream);
+                }
+            }
+        }
+    }
+
+    pub fn flush(&mut self) {
+        self.activate_slices.clear();
+    }
+}
+
+pub type Binding = cubecl_hip_sys::hipDeviceptr_t;
+
+/// The memory resource that can be allocated for the device.
+#[derive(new, Debug)]
+pub struct HipResource {
+    /// The buffer.
+    pub ptr: cubecl_hip_sys::hipDeviceptr_t,
+    pub binding: Binding,
+    offset: u64,
+    size: u64,
+}
+
+unsafe impl Send for HipResource {}
+
+impl HipResource {
+    /// Return the binding view of the buffer.
+    pub fn as_binding(&self) -> Binding {
+        self.binding
+    }
+
+    /// Return the buffer size.
+    pub fn size(&self) -> u64 {
+        self.size
+    }
+
+    /// Return the buffer offset.
+    pub fn offset(&self) -> u64 {
+        self.offset
+    }
```

Comment:
Those two could simply become pub fields. Before we actually add something to do in those functions, but not anymore.